feat: ability to assign people to round robin from org members

Amit91848 opened the pull request:
## What does this PR do?

Fixes #14190 


https://github.com/calcom/cal.com/assets/74371312/84a92e94-cfa7-4e7f-b4c3-3cbb10a598fd



## Type of change

- New feature (non-breaking change which adds functionality)


## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

Amit91848 committed:
feat: ability to assign people to round robin from org members

Amit91848 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat/assign_to_round_robin_from_org_members

Amit91848 committed:
fix: include schedulingtype in filter

Amit91848 committed:
Merge branch 'main' into feat/assign_to_round_robin_from_org_members

Amit91848 committed:
fix: type check

Amit91848 committed:
Merge branch 'main' into feat/assign_to_round_robin_from_org_members

Amit91848 committed:
Merge branch 'main' into feat/assign_to_round_robin_from_org_members

ibex088 committed:
Merge branch 'main' into feat/assign_to_round_robin_from_org_members

supalarry commented on packages/features/users/components/UserTable/BulkActions/EventList.tsx:
```diff
@@ -0,0 +1,193 @@
+                    const isSelected = ids.every((id) => selectedEvents.has(id));
+                    return (
+                      <>
+                        <ListItem
```
Component on line 93 renders team name - what if the team has no event types ? There is no point in rendering it then, so we should render team name component only if `events.length > 0`.

supalarry commented on packages/features/users/components/UserTable/BulkActions/EventList.tsx:
```diff
@@ -0,0 +1,193 @@
+      showToast(error.message, "error");
+    },
+    onSuccess: (res) => {
+      showToast(`${res.count} Users added to ${Array.from(selectedEvents).length} events`, "success");
```
`res.count` produces incorrect result - if you add 1 user to 2 round robin events:

<img width="959" alt="Screenshot 2024-05-15 at 11 23 07" src="https://github.com/calcom/cal.com/assets/42170848/767e5ca8-a8ce-4407-bf8c-d5b5e8ddb673">

Then notification on bottom right corner says 2 users added:
<img width="1140" alt="Screenshot 2024-05-15 at 11 23 11" src="https://github.com/calcom/cal.com/assets/42170848/8243ad6e-fb49-4516-abc2-4f0223318ea9">

Amit91848 replied:
Yes actually it return the number of host connections made, so ig a more correct message would be 2 Hosts added to 2 events. 

Or should we show number of users only?

supalarry replied:
If you were a user, which one would be clearer for you? Personally, I think that `X users added to Y events` is the most straight-forward. If I added 1 user to 2 events and saw that `2 hosts added to 2 events` I would be bit confused.

Amit91848 replied:
I agree hosts could be more confusing. Changed it to users. 

Although now if user is already a host for the event type the message will show as added 1 user, even if no changes were made.

supalarry commented on packages/features/users/components/UserTable/BulkActions/EventList.tsx:
```diff
@@ -0,0 +1,193 @@
+              <CommandEmpty>No results found.</CommandEmpty>
+              <CommandGroup>
+                {eventTypeGroups &&
+                  eventTypeGroups.map((team) => {
```
Can we rename `eventTypeGroups` constant to `teams` ? When mapping the individual item is a team really.

supalarry commented on packages/features/users/components/UserTable/BulkActions/EventList.tsx:
```diff
@@ -0,0 +1,193 @@
+                  eventTypeGroups.map((team) => {
+                    const events = team.eventTypes;
+                    const ids = events.map((event) => event.id);
+                    const teamId = team.teamId || 0;
```
I would remove `|| 0` and render `ListItem` only if `teamId` is defined - code should not render anything if the `teamId` is not given because if it did, then something is broken, and that data would be sent to backend - the logic should only work if `teamId` is defined.

supalarry commented on packages/features/users/components/UserTable/BulkActions/EventList.tsx:
```diff
@@ -0,0 +1,193 @@
+  const utils = trpc.useUtils();
+  const teamIds = teams?.map((team) => team.id);
+  const { data } = trpc.viewer.eventTypes.getByViewer.useQuery({
+    filters: { teamIds, schedulingType: SchedulingType.ROUND_ROBIN },
```
@PeerRich you created the issue - why are we requesting ability to assign people from org members specifically to round robin? Why not also collective?

We call this component `EventList` but really it's `RoundRobinList` - to me it makes sense to also include collective event types not only round robin.

supalarry commented on packages/lib/event-types/getEventTypesByViewer.ts:
```diff
@@ -30,6 +30,7 @@ type User = {
 type Filters = {
   teamIds?: number[];
   upIds?: string[];
+  schedulingType?: SchedulingType;
```
What if we want to fetch round robin and collective? I think `schedulingType` should be an array.

supalarry commented on packages/trpc/server/routers/viewer/organizations/utils.ts:
```diff
@@ -0,0 +1,83 @@
+  input: TAddBulkTeams;
+}
+
+export const addBulkToTeam = async ({ user, input }: AddBulkToTeamProps) => {
```
Could this function have a better name? It's called `addBulkToTeam` but it's not really clear what `Bulk` is.

supalarry commented on packages/lib/event-types/getEventTypesByViewer.ts:
```diff
@@ -209,6 +210,13 @@ export const getEventTypesByViewer = async (user: User, filters?: Filters, forRo
     }
     return filters?.teamIds?.includes(eventType?.teamId || 0) ?? false;
   };
+  const filterSchedulingTypeEventTypesBasedOnInput = (evType: Awaited<ReturnType<typeof mapEventType>>) => {
```
can simply call it `applyInputFilters` or something

supalarry reviewed (changes requested):
Good work Amit! Requested some changes. Let me know what you think! : )

supalarry commented:
Also, I would change the icon of "Add to event type":

<img width="959" alt="Screenshot 2024-05-15 at 11 23 07" src="https://github.com/calcom/cal.com/assets/42170848/6d1fd850-2e91-4e8f-bcd3-e79054c5d5ab">

I think it makes sense to have same icon as "Event types" at the main page of dashboard sidebar:
<img width="197" alt="Screenshot 2024-05-15 at 10 39 10" src="https://github.com/calcom/cal.com/assets/42170848/63cbfb35-ea19-4ad3-be43-ca3b89af15b9">

Amit91848 commented:
> Good work Amit! Requested some changes. Let me know what you think! : )

Will look into it.

Amit91848 committed:
rename to EventTypesList

Amit91848 committed:
rename to teams, change icon to link, early return if no events or teamId

Amit91848 committed:
filter scheduling type array, change filter name

Amit91848 committed:
rename to bulkAddMembersToTeams and bulkAddMembersToEventTypes

supalarry commented on packages/lib/event-types/getEventTypesByViewer.ts:
```diff
@@ -203,12 +204,24 @@ export const getEventTypesByViewer = async (user: User, filters?: Filters, forRo
+
+    if (!filters.schedulingType) return true;
+
+    if (!evType.schedulingType) return false;
```
Question: What is the logic for returning false if event type has no schedulingType ?

If we remove `if (!evType.schedulingType) return false;` from code, then code below:
```
return filters.schedulingType.includes(evType.schedulingType);
```

would perform the same logic, right? So I think that `if (!evType.schedulingType) return false;` can be removed.

Amit91848 replied:
Nvm fixed this

supalarry commented on packages/features/users/components/UserTable/BulkActions/EventTypesList.tsx:
```diff
@@ -36,7 +36,7 @@ export function EventTypesList({ table, orgTeams }: Props) {
     filters: { teamIds, schedulingType: [SchedulingType.ROUND_ROBIN] },
   });
-  const mutation = trpc.viewer.organizations.bulkAddToEventType.useMutation({
+  const mutation = trpc.viewer.organizations.bulkAddMembersToEventTypes.useMutation({
```
Now that I think, I think the most straight forward name to this mutation would be `addMembersToEventTypes` without `bulk` in the name. And then also renaming `addBulkMembersToEventTypes.handler` to `addMembersToEventTypes.handler` - what do you think?

Amit91848 replied:
Yeah makes sense

Amit91848 replied:
While I am at it should I refactor for bulkDeleteUsers as well?

supalarry replied:
I think it's best to create a new PR for "bulkDeleteUsers" as it falls out of the scope of this task : )

supalarry reviewed (commented):
Great refactor! : )

I replied to your comment in [PR comment](https://github.com/calcom/cal.com/pull/14199#discussion_r1601291501) and here are couple more minor refactors below. Let me know what you think!

Amit91848 committed:
filter refactor

Amit91848 committed:
remove bulk

Amit91848 committed:
Merge branch 'main' into feat/assign_to_round_robin_from_org_members

Amit91848 committed:
filter remove null return

Amit91848 commented on packages/lib/event-types/getEventTypesByViewer.ts:
```diff
@@ -31,7 +31,7 @@ type User = {
   teamIds?: number[];
   upIds?: string[];
-  schedulingTypes?: SchedulingType[];
+  schedulingTypes?: (SchedulingType | null)[];
```
> Yes would be the same but I did it because I couldn't shake the type error without early return.
> 
> ![Screenshot 2024-05-16 at 1 49 39 PM](https://private-user-images.githubusercontent.com/74371312/331113506-71694bfb-f78a-4729-899a-2180a81af3ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NTI1MjQsIm5iZiI6MTcxNTg1MjIyNCwicGF0aCI6Ii83NDM3MTMxMi8zMzExMTM1MDYtNzE2OTRiZmItZjc4YS00NzI5LTg5OWEtMjE4MGE4MWFmM2VmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDA5MzcwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmNWU3MDM0OTkwNmU1NjFmMGM5ZDUyMzA2NjlkMTMwNzA2ZmM4ODIyZjIwMTUxMTlhM2U0ZjgzOWQ0ZGRjMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qeM16jgWKHTP4bETl5eGXWNP4ftgktJKKByT7zhlbCo)


Then we get this type error

Amit91848 committed:
Revert "filter remove null return"

This reverts commit 8ae6296dac24e5081b5896bb9dfe0c000c0f7faf.

supalarry reviewed (commented):
If a user is already part of round robin event-type `rocky-rr`

<img width="1493" alt="Screenshot 2024-05-17 at 10 54 00" src="https://github.com/calcom/cal.com/assets/42170848/d0881918-3b59-4b2e-8c3c-c1608377c330">


and we select it and click "Add to event type", then the `rocky-rr` is unchecked even though the user is part of it, which can confuse users because they don't know the state of whether or not a user already is part of that event type:

<img width="689" alt="Screenshot 2024-05-17 at 10 56 03" src="https://github.com/calcom/cal.com/assets/42170848/d0b1a6dd-9705-4731-b41c-19c3e86d54aa">

Do you think you can adjust this - if a user is part of the event-type, then have it checked after clicking "Add to event type":
<img width="652" alt="Screenshot 2024-05-17 at 10 58 31" src="https://github.com/calcom/cal.com/assets/42170848/e66914df-4d26-42f0-ae7e-0945531da0f2">



Thank you!

---

cc @sean-brydon the same happens when clicking "Add to team" - even though the user is already part of the team, the checkboxes are unchecked.

<img width="402" alt="Screenshot 2024-05-17 at 11 00 42" src="https://github.com/calcom/cal.com/assets/42170848/65d8f755-ef75-4085-8674-73e2e0824ee8">

Amit91848 committed:
show checked if user already host in event type

Amit91848 committed:
check boxes if user is part of the team

Amit91848 commented:
> Do you think you can adjust this - if a user is part of the event-type, then have it checked after clicking "Add to event type":
Added this. Currently clicking the boxes does nothing.

> cc @sean-brydon the same happens when clicking "Add to team" - even though the user is already part of the team, the checkboxes are unchecked.

Since it hardly required any changes, fixed it for teams here itself.

https://github.com/calcom/cal.com/assets/74371312/bb1b5301-f457-431e-8738-993bb1eaeaa5

supalarry commented:
> > Do you think you can adjust this - if a user is part of the event-type, then have it checked after clicking "Add to event type":
> > Fair enough, just wanted to confirm a few things.
> 
> Suppose the events where user is a host is checked, does clicking on them uncheck them, will anything happen on checking them? Or should we just disable those?
> 
> Understandably if multiple users are selected only check those where all the users are a part of.

I would expect that unchecking would do nothing by itself, but unchecking and then clicking "Apply" should remove them from the round robin.

supalarry reviewed (commented):
Amazing to see that event-types user is already part of it checked, and thanks a lot for doing it for teams too!

Do you think it's possible that if user is part of the event-type or team, to enable un-checking the box and after pressing "Apply" the user would be removed from the event-type or team?

Amit91848 commented:
> I would expect that unchecking would do nothing by itself, but unchecking and then clicking "Apply" should remove them from the round robin.

That's what I thought too.

Amit91848 committed:
remove hosts on unchecking boxes

Amit91848 commented:
> Do you think it's possible that if user is part of the event-type or team, to enable un-checking the box and after pressing "Apply" the user would be removed from the event-type or team?

https://github.com/calcom/cal.com/assets/74371312/f0d8f033-1a1b-4ebf-af39-1455981147ff

Added this for event types. 
For teams should I change removeMember handler to accept array of teams and members or create a new handler?

https://github.com/calcom/cal.com/blob/98235d00c8177c97bd55b738b23028271824f52f/packages/trpc/server/routers/viewer/teams/removeMember.schema.ts#L3-L7

Amit91848 commented on packages/features/users/components/UserTable/BulkActions/EventTypesList.tsx:
```diff
@@ -0,0 +1,262 @@
+                              isTeam={false}
+                              onSelect={() => {
+                                if (!isSelected) {
+                                  if (areAllUsersHostForEventType) {
```
If there are 3 events A, B, C and user is part of event B. If you click on team, all events gets selected. If you uncheck team, all events get unchecked. Mind you we have not clicked apply yet so user is still a host of event B, but now if you click Event B nothing happens (it should check event b).

Maybe I could name this better lmk what you think

Amit91848 commented on packages/features/users/components/UserTable/BulkActions/EventTypesList.tsx:
```diff
@@ -0,0 +1,262 @@
+                                    addValue(selectedTeams, setSelectedTeams, [teamId]);
+                                  }
+                                } else {
+                                  if (areAllUsersHostForEventType) {
```
If checked and user is already a host, then we have to remove from that event

Amit91848 committed:
fix isSelected condition

Amit91848 commented on packages/features/users/components/UserTable/BulkActions/EventTypesList.tsx:
```diff
@@ -0,0 +1,262 @@
+                              isTeam={false}
+                              onSelect={() => {
+                                if (!isSelected) {
+                                  if (areAllUsersHostForEventType) {
```
If all selected users are host and it is still unchecked, then it was because it was added to removeEvents, so remove it from there

supalarry reviewed (dismissed):
Thanks for all the refactors and it's an amazing PR! Thank you! : )

Tested locally and adding and removing members from event types work as expected! 🚀

Amit91848 commented:
Glad to see you liked it @supalarry . 
I think this is still incomplete, I haven't implemented removing members from teams yet, if you can confirm this I will get to working on it too.

> > Do you think it's possible that if user is part of the event-type or team, to enable un-checking the box and after pressing "Apply" the user would be removed from the event-type or team?
>
>  Added this for event types. For teams should I change removeMember handler to accept array of teams and members or create a new handler?
> 
> https://github.com/calcom/cal.com/blob/98235d00c8177c97bd55b738b23028271824f52f/packages/trpc/server/routers/viewer/teams/removeMember.schema.ts#L3-L7

supalarry commented:
> Glad to see you liked it @supalarry . I think this is still incomplete, I haven't implemented removing members from teams yet, if you can confirm this I will get to working on it too.
> 
> > > Do you think it's possible that if user is part of the event-type or team, to enable un-checking the box and after pressing "Apply" the user would be removed from the event-type or team?
> > 
> > 
> > Added this for event types. For teams should I change removeMember handler to accept array of teams and members or create a new handler?
> > https://github.com/calcom/cal.com/blob/98235d00c8177c97bd55b738b23028271824f52f/packages/trpc/server/routers/viewer/teams/removeMember.schema.ts#L3-L7

That would be amazing thank you!

Amit91848 committed:
remove selected members from teams

Amit91848 commented:
@supalarry please review again


https://github.com/calcom/cal.com/assets/74371312/bb04ded5-61a7-4508-9fa2-477257c6e24b

supalarry reviewed (approved):
Amazing work! Thank you for the refactors and going extra mile for updating teams management!

PeerRich commented:
how does this look on mobile?
